modules/smtp/functions.php

```diff
@@ -61,3 +61,92 @@ function get_reply_type($request) {
         return false;
     }
 }
+
+/**
+ * @subpackage smtp/functions
+ */
+if (!hm_exists('send_scheduled_message')) {
+function send_scheduled_message($handler, $imap, $folder, $msg_id, $send_now = false) {    
+    $msg_headers = $imap->get_message_headers($folder, $msg_id);    
+    $imap_details = $imap->get_config();       
+
+    try {
+        if (empty($msg_headers['X-Schedule'])) {
+            return false;
+        }
+
+        if (new DateTime($msg_headers['X-Schedule']) <= new DateTime() || $send_now) {
+            $profile = Hm_Profiles::get($msg_headers['X-Profile-ID']);
+            if (!$profile) {
+                $profiles = Hm_Profiles::search('server', $imap_details['server']);
+
+                if (!$profiles) {
+                    Hm_Debug::add(sprintf('ERRCannot find profiles corresponding with IMAP server: %s', $imap_details['server']));
+                    return false;
+                }
+                $profile = $profiles[0];
+            }
+
+            $smtp = Hm_SMTP_List::connect($profile['smtp_id'], false);
+
+            if ($smtp && $smtp->authed()) {
+                $delivery_receipt = isset($msg_headers['X-Delivery']);
+
+                $recipients = [];
+                foreach (['To', 'Cc', 'Bcc'] as $fld) {
+                    if (array_key_exists($fld, $msg_headers)) {
+                        $recipients = array_merge($recipients, Hm_MIME_Msg::find_addresses($msg_headers[$fld]));
+                    }
+                }
+
+                $msg_content = $imap->get_message_content($folder, $msg_id, 0);
+                $from = process_address_fld($msg_headers['From']);
+
+                $err_msg = $smtp->send_message($from[0]['email'], $recipients, $msg_content, $delivery_receipt);
+
+                if (!$err_msg) {
+                    $imap->delete_message($folder, $msg_id, false);
+                    save_sent_msg($handler, $imap->get_config()['id'], $imap, $imap_details, $msg_content, $msg_id, false);
+                    return true; 
+                }
+            }
+        }
+    } catch (Exception $e) {
+        Hm_Debug::add(sprintf('ERRCannot send message: %s', $msg_headers['subject']));
+    }
+    return false; 
+}}
+
+/**
+ * @subpackage smtp/functions
+ */
+if (!hm_exists('reschedule_message_sending')) {
+function reschedule_message_sending($handler, $imap, $msg_id, $folder, $new_date) {
```

Comment:
This should receive a Mailbox object, not imap.

Reply:
Thank you @kroky. I just need to rename it but the object is of type Hm_Mailbox if you see the way it is used you will notice that.